[pip][design] PIP-281: Add notifyError method on PushSource (#20807)
apache/pulsar: pip/pip-281.md

# Title: [io] Add notifyError method on PushSource

## Motivation

> In function framework, when [source.read()](https://github.com/apache/pulsar/blob/f7c0b3c49c9ad8c28d0b00aa30d727850eb8bc04/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L496-L506) method throw exception, it will trigger close function instance. If it is in the k8s environment, it will be restarted

Reviewer: ```suggestion
In function framework, when [source.read()](https://github.com/apache/pulsar/blob/f7c0b3c49c9ad8c28d0b00aa30d727850eb8bc04/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L496-L506) method throws an exception, it will trigger close function instance. If it is in the k8s environment, it will be restarted
```

> On io source connector, we provide [PushSource](https://github.com/apache/pulsar/blob/branch-3.0/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/PushSource.java) class for users to use, and users can extend this class to quickly implement the push message model.

Reviewer: ```suggestion
on IO source connector, you can use the [PushSource](https://github.com/apache/pulsar/blob/branch-3.0/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/PushSource.java) class and extend it to quickly implement the push message model.
```

It overrides the `read` method and provides the `consume` method for the user to call.

> However, if the source connector that extends from the class, 

Reviewer: ```suggestion
However, if the source connector extends from the class, 
```

it cannot notify the function framework if it encounters an exception while consuming data internally, 
in other words, the function call `source.read()` never triggers an exception and never exits the process.

## Goals

> Add `notifyError` method on PushSource, This method can receive an exception and put the exception in the queue. The next time an exception is `read`, will throws it.

Reviewer: ```suggestion
Add the `notifyError` method on PushSource. This method can receive an exception and put the exception in the queue. The next time an exception is `read`, it will throw it.
```

throw "it", throw what?

```java

  public Record<T> read() throws Exception {
    Record<T> record = queue.take();
    if (record instanceof ErrorNotifierRecord) {
      throw ((ErrorNotifierRecord) record).getException();
    }
    return record;
  }

  /**
   * Allows the source to notify errors asynchronously.
   * @param ex
   */
  public void notifyError(Exception ex) {
    consume(new ErrorNotifierRecord(ex));
  }
}
```

Just like the implementation of the current [BatchPushSource](https://github.com/apache/pulsar/blob/branch-3.0/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/BatchPushSource.java)

### Compatibility

> This PIP is to provide a method for users to use, not an new interface. 

Reviewer: ```suggestion
This PIP is to provide a method for users rather than introducing a new interface. 
```

- So it is forward compatible
- However, connector using this method are not backward compatible.
For example, the currently Kafka source connector compiles with version 3.1(include this pip) Pulsar dependencies and uses the `notifyError` method, 

> if it switches back to version 3.0(exclude this pip) Pulsar to compile, it will encounter compile errors.

Reviewer: 1. What's the meaning of "(include this pip)"?
2. Does the Kafka source connector "use" a method?
3. "Who" switches back to 3.0? Do you mean Kafka source connector complies with Pulsar  3.0?

Author: > What's the meaning of "(include this pip)"?

That means the 3.1 version has this pip-related feature.

> Does the Kafka source connector "use" a method?

This is an example, in fact it is not currently used, and then this PIP will be PR to fix it

> "Who" switches back to 3.0? Do you mean the Kafka source connector complies with Pulsar 3.0?
Kafka source connector. No, I'm giving examples of incompatible scenarios.

hi @Anonymitaet Please help take another look at the latest sentence, thanks.

> For example, the Kafka source connector compiles with version 3.1(include this pip) Pulsar dependencies and uses the `notifyError` method, and if it switches back to version 3.0(exclude this pip) Pulsar compilation, it will encounter compilation errors.

Reviewer: This is a long sentence with unclear clauses. What do you intend to mean?

Author: I refactor this sentence, PTAL.

Reviewer: Sorry, it's still unclear. What's the relationship between "the Kafka source connector compiles with version 3.1(include this pip) Pulsar dependencies and uses the `notifyError` method," and the latter sentence?

### In Scope

> Use this method, current source connectors can extends the `PushSource`, and use `notifyError` method to throw exception. Such as:

Reviewer: "Use this method", what does "this" refer to? Need to show it explicitly

> Use this method, Like all current source connectors that extends the PushSource, process exit can be implemented. Such as:

Reviewer: This sentence is unclear. What do you intend to mean?

Author: I refactor this sentence, PTAL.

it cannot notify the function framework if it encounters an exception while consuming data internally, 
in other words, the function call `source.read()` never triggers an exception and never exits the process.

## Goals

Add `notifyError` method on PushSource, This method can receive an exception and put the exception in the queue. The next time an exception is `read`, will throws it.
```java

  public Record<T> read() throws Exception {
    Record<T> record = queue.take();
    if (record instanceof ErrorNotifierRecord) {
      throw ((ErrorNotifierRecord) record).getException();
    }
    return record;
  }

  /**
   * Allows the source to notify errors asynchronously.
   * @param ex
   */
  public void notifyError(Exception ex) {
    consume(new ErrorNotifierRecord(ex));
  }
}
```

Just like the implementation of the current [BatchPushSource](https://github.com/apache/pulsar/blob/branch-3.0/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/BatchPushSource.java)

BTW: This is a very simple change and is forward compatible. Sorry, I didn't notice that this change requires PIP before, so the related PRs have been merged.
- https://github.com/apache/pulsar/pull/20791

> If this PIP vote does not pass, I revert this PR after that.

Reviewer 2: I think we could remove these line:
> Sorry, I didn't notice that this change requires PIP before, so the related PRs have been merged.
> - https://github.com/apache/pulsar/pull/20791
> If this PIP vote does not pass, I revert this PR after that.

It's not much relevant to this PIP.

Author: Agree, just comment on this PR.

### Compatibility

> This PIP is to provide a method for users to use, not an new interface. 

Reviewer 2: I think the `new interface` here is referred to the Java interface? But adding a new method to a class is actually adding a new interface method(although it's not a Java interface class). Right?

Author: > I think the new interface here is referred to the Java interface?

Yes, I mean, not a new interface method, so don't have to force the user to implement it. So it is forward compatible

Reviewer 2: Thanks. I get your point

This PIP is to provide a method for users rather than introducing a new interface.

- So it is forward compatible

> - However, connector using this method are not backward compatible. 
> For example, If `Kafka source` connector depends on version v3.0(include this pip feature) of pulsar-io and uses the `notifyError` method, 
> when it switches back to depends on 3.0(not include this pip feature) version of pulsar-io to compile will encounter compile errors.

Reviewer: ```suggestion
- However, connectors using this method are not backward compatible. 
For example, If a Kafka source connector built upon pulsar-io v3.1 (including features introduced in this PIP) and uses the `notifyError` method, 
when it switches back to pulsar-io v3.0 (excluding features introduced in this PIP), it will encounter errors during compilation. 
```

do you intend to mean this?
